utils/s3/client.cc

```diff
@@ -854,6 +861,209 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
+            }
+            s3l.trace("created multipart upload for {} -> id = {}", _object_name, _upload_id);
+        });
+    }
+
+    future<> complete_multipart_upload() {
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html
+        unsigned parts_xml_len = prepare_multipart_upload_parts(_part_etags);
+        auto req = http::request::make("POST", _client->_host, _object_name);
+        req.query_parameters["uploadId"] = _upload_id;
+        req.write_body("xml", parts_xml_len, [this] (output_stream<char>&& out) -> future<> {
+            return dump_multipart_upload_parts(std::move(out), _part_etags);
+        });
+        // If this request fails, complete_multipart_upload() throws, the upload should then
+        // be aborted in .close() method
+        co_await _client->make_request(std::move(req));
+        _upload_id = ""; // now upload_started() returns false
+    }
+
+    future<> abort_multipart_upload() {
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_AbortMultipartUpload.html
+        s3l.trace("abort multipart upload {}", _upload_id);
+        auto req = http::request::make("DELETE", _client->_host, _object_name);
+        req.query_parameters["uploadId"] = std::exchange(_upload_id, "");
+        co_await _client->make_request(std::move(req), ignore_reply, http::reply::status_type::no_content);
+    }
+
+    class throttled_consumer {
+        output_stream<char>& _output;
+        client& _client;
+        using consumption_result_type = consumption_result<char>;
+
+    public:
+        throttled_consumer(output_stream<char>& output, client& client)
+            : _output{output}
+            , _client{client} {
+        }
+        future<consumption_result_type> operator()(temporary_buffer<char> buf) {
+            if (buf.empty()) {
+                co_return stop_consuming(std::move(buf));
+            }
+            auto mem_units = co_await _client.claim_memory(buf.size());
+            co_await _output.write(buf.get(), buf.size());
+            co_return continue_consuming();
+        }
+    };
+
+    future<> upload_part(file&& f, uint64_t offset, uint64_t part_size) {
+        // upload a part in a multipart upload, see
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+        unsigned part_number = _part_etags.size();
+        _part_etags.emplace_back();
+        auto req = http::request::make("PUT", _client->_host, _object_name);
+        req._headers["Content-Length"] = to_sstring(part_size);
+        req.query_parameters.emplace("partNumber", to_sstring(part_number + 1));
+        req.query_parameters.emplace("uploadId", _upload_id);
+        s3l.trace("PUT part {}, {} bytes (upload id {})", part_number, part_size, _upload_id);
+        req.write_body("bin", part_size, [f=std::move(f), offset, part_size, this] (output_stream<char>&& out_) mutable -> future<> {
+            auto output = std::move(out_);
```

Comment:
Worth generalizing this body for upload_part()'s request and put_object()'s one, the only difference is offset which can be 0 for put_object()

Reply:
but there are more of them: 

- pass "partNumber" and "uploadId" as query parameter for multi part upload
- send the optional tag in request for put object
- collect the `ETag` returned in the response for multi part upload

do you still think that we should generalize it?

Reply:
> but there are more of them:
> 
>     * pass "partNumber" and "uploadId" as query parameter for multi part upload
>     * send the optional tag in request for put object
>     * collect the `ETag` returned in the response for multi part upload
> 
> 
> do you still think that we should generalize it?

Wait, wait, I'm talking about req.write_body() lambdas, there's neither partNumber nor uploadId in there

Reply:
i see. thanks. much better and clean now.